GUI/Types/Renderer/Shaders/simple.frag

```diff
@@ -327,60 +418,91 @@ void main()
 
     if (visibility > 0.0)
     {
-        Lo += specularContribution(L, V, N, F0, albedo, metalness, roughness) * visibility;
-        Lo += diffuseLobe(max(dot(N, L), 0.0) * getSunColor()) * visibility;
+        vec3 specularLight = specularLighting(L, V, N, F0, specularColor, roughness, extraParams);
+        float diffuseLight = diffuseLobe(max(dot(N, L), 0.0), roughness);
+
+       lighting.SpecularDirect += specularLight * visibility * getSunColor();
+       lighting.DiffuseDirect += diffuseLight * visibility * getSunColor();
     }
 
+
+    // Indirect Lighting
     #if (D_BAKED_LIGHTING_FROM_LIGHTMAP == 1) && (LightmapGameVersionNumber > 0)
         irradiance = texture(g_tIrradiance, vLightmapUVScaled).rgb;
         vec4 vAHDData = texture(g_tDirectionalIrradiance, vLightmapUVScaled);
-        const float DirectionalLightmapMinZ = 0.05;
-        irradiance *= mix(1.0, vAHDData.z, DirectionalLightmapMinZ);
-        occlusion *= vAHDData.w;
+
+        irradiance = ComputeLightmapShading(irradiance, vAHDData, normal);
+
+        // Lightmap AO is only used in Lightmap Version 2
+        #if (LightmapGameVersionNumber == 2)
+            occlusion = min(occlusion, vAHDData.w);
+        #endif
     #elif (D_BAKED_LIGHTING_FROM_VERTEX_STREAM == 1)
         irradiance = vPerVertexLightingOut.rgb;
     #endif
 
-    Lo *= occlusion;
 
-    outputColor.rgb *= (irradiance);
-    outputColor.rgb += Lo;
+    lighting.DiffuseIndirect = irradiance;
 
     // Environment Map
     #if (S_SPECULAR == 1)
-        vec3 specular = GetEnvironment(N, V, roughness, F0, irradiance);
-        outputColor.rgb += specular * occlusion;
+        vec3 specular = GetEnvironment(N, V, roughness, specularColor, irradiance, extraParams);
+        lighting.SpecularIndirect += specular * occlusion;
     #endif
 
+    lighting.DiffuseDirect *= mix(1.0, occlusion, g_flAmbientOcclusionDirectDiffuse);
+    lighting.SpecularDirect *= mix(1.0, occlusion, g_flAmbientOcclusionDirectSpecular);
+
+    vec3 diffuseLighting = lighting.DiffuseDirect + lighting.DiffuseIndirect * occlusion;
+    vec3 specularLighting = lighting.SpecularDirect + lighting.SpecularIndirect * occlusion;
+
+    #if F_NO_SPECULAR_AT_FULL_ROUGHNESS == 1
+        specularLighting = (roughness == 1.0) ? vec3(0) : specularLighting;
+    #endif
 
+    vec3 combinedLighting = diffuseColor * diffuseLighting + specularLighting;
+
+    outputColor.rgb = combinedLighting;
+
+#endif
+#if F_DISABLE_TONE_MAPPING == 0
     outputColor.rgb = pow(outputColor.rgb, vec3(invGamma));
     //outputColor.rgb = SRGBtoLinear(outputColor.rgb);
 #endif
```

Comment:
Shouldn't it be inside the #endif? I imagine unlit materials don't not need any correction.

Reply:
Funnily enough, it's actually the reverse: vr_complex only exposes this parameter if the material is unlit.